astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+ public:
+  static constexpr TimeOfDay hhmmss_ns(int const hhmmss, int ns);
+
+  constexpr int hour() const;
+  constexpr int minute() const;
+  constexpr int second() const;
+  constexpr int nanosecond() const;
+
+  constexpr bool is_leap_second() const;
+  // Whether |*this| is 24:00:00.
+  constexpr bool is_end_of_day() const;
+
+ private:
+  constexpr TimeOfDay(int const hour,
+                      int const minute,
+                      int const second,
+                      int const nanosecond);
+
+  // Checks that this represents a valid time of day as per ISO 8601, thus
+  // that the components are in the normal range, or that the object represents
+  // a time in a leap second, or that it represents the end of the day.
+  constexpr TimeOfDay const& checked() const;
+
+  int const hour_;
+  int const minute_;
+  int const second_;
+  int const nanosecond_;
+};
+
+class DateTime {
+ public:
+  constexpr Date const& date() const;
+  constexpr TimeOfDay const& time() const;
+
+  // If |time()| is 24:00:00, returns an equivalent DateTime where midnight is
+  // expressed as 00:00:00 on the next day; otherwise, returns *this.
+  constexpr DateTime normalized_end_of_day() const;
+
+ private:
+  constexpr DateTime(Date const date, TimeOfDay const time);
+
+  // Checks that |time| does not represent a leap second unless |date| is the
+  // last day of June, December, March, or September.
+  constexpr DateTime const& checked() const;
+
+  Date const date_;
+  TimeOfDay const time_;
+
+  friend constexpr DateTime operator""_DateTime(char const* string,
+                                                 std::size_t size);
+};
+
+// Arithmetico-calendrical utility functions.
+
+constexpr std::array<int, 12> non_leap_year_month_lengths{
+    {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}};
+
+constexpr bool is_gregorian_leap_year(int const year) {
+  return (year % 4 == 0 && (year % 100 != 0 || year % 400 == 0));
+}
+
+constexpr int gregorian_year_length(int const year) {
+  return is_gregorian_leap_year(year) ? 366 : 365;
+}
+
+constexpr int month_length(int year, int month) {
+  return (is_gregorian_leap_year(year) && month == 2)
+             ? 29
+             : non_leap_year_month_lengths[month - 1];
+}
+
+constexpr int mod_7_offset_1 (int const x) {
+  return x % 7 == 0 ? 7 : (x % 7);
+}
+
+// Result in [1, 7], 1 is Monday.
+constexpr int day_of_week_on_january_1st(int const year) {
+  // Gauss's formula, see
+  // https://en.wikipedia.org/wiki/Determination_of_the_day_of_the_week#Gauss.27s_algorithm.
+  return mod_7_offset_1(1 +
+                        5 * ((year - 1) % 4) +
+                        4 * ((year - 1) % 100) +
+                        6 * ((year - 1) % 400));
+}
+
+constexpr bool is_53_week_year(int const year) {
+  return day_of_week_on_january_1st(year) == 4 ||
+         (is_gregorian_leap_year(year) &&
+          day_of_week_on_january_1st(year) == 3);
+}
+
+constexpr int ordinal_of_w_01_1(int const year) {
```

Comment:
Done:
```cpp
// Returns the ordinal in |year| of the first day of the first week of |year|.
// The result is in [-2, 4], with values in [-2, 0] meaning that the first week
// of |year| starts in |year - 1|.
// A result in [-2, 1] means that the first day of |year| is in the first week
// of |year|; otherwise, it is in the last week of |year - 1|.
```
It so happens that this is actually equal to
```mathematica
Mod[2 - d, 7, -2]
```
where `d` is `day_of_week_on_january_1st(year) `, should it be rewritten that way?
The current formulation separates the case where January 1st of `year` is in a week of `year` from the case where it is not.